packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/FunctionsTests.kt

```diff
@@ -0,0 +1,648 @@
+                    val nowAsDuration: Duration = RealmInstant.now().toDuration()
+                    val nowInMilliseconds = nowAsDuration.inWholeMilliseconds.milliseconds
+                    val now = nowInMilliseconds.toRealmInstant()
+                    assertEquals(
+                        now,
+                        functions.callBlocking<RealmInstant>(FIRST_ARG_FUNCTION.name, now)
+                    )
+
+                    BsonDateTime().let {
+                        assertEquals(it, functions.callBlocking(FIRST_ARG_FUNCTION.name, it))
+                    }
+                }
+                BsonType.UNDEFINED -> assertEquals(
+                    BsonUndefined,
+                    functions.callBlocking(FIRST_ARG_FUNCTION.name, BsonUndefined)
+                )
+                BsonType.NULL -> {
+                    assertEquals(
+                        BsonNull,
+                        functions.callBlocking(FIRST_ARG_FUNCTION.name, BsonNull)
+                    )
+                    assertNull(functions.callBlocking(FIRST_ARG_FUNCTION.name, null))
+                }
+                BsonType.REGULAR_EXPRESSION -> assertTypeOfFirstArgFunction(BsonRegularExpression(""))
+                BsonType.SYMBOL -> assertTypeOfFirstArgFunction(BsonSymbol(""))
+                BsonType.JAVASCRIPT -> assertTypeOfFirstArgFunction(BsonJavaScript(""))
+                BsonType.JAVASCRIPT_WITH_SCOPE -> assertTypeOfFirstArgFunction(
+                    BsonJavaScriptWithScope("", BsonDocument())
+                )
+                BsonType.TIMESTAMP -> assertTypeOfFirstArgFunction(BsonTimestamp())
+                BsonType.MIN_KEY -> assertTypeOfFirstArgFunction(BsonMinKey)
+                BsonType.MAX_KEY -> assertTypeOfFirstArgFunction(BsonMaxKey)
+                BsonType.DB_POINTER -> assertTypeOfFirstArgFunction(
+                    BsonDBPointer(
+                        namespace = "namespace",
+                        id = BsonObjectId()
+                    )
+                )
+                BsonType.END_OF_DOCUMENT -> {
+                    // Not a real Bson type
+                }
+                else -> {
+                    fail("Unsupported BsonType $type")
+                }
+            }
+        }
+    }
+
+    private inline fun <reified T : Any> assertTypeOfFirstArgFunction(
+        value: T
+    ): T = functions.callBlocking<T>(FIRST_ARG_FUNCTION.name, value).also {
+        assertEquals(value, it)
+    }
+
+//    @Test
+//    fun asyncCallFunction() = looperThread.runBlocking {
+//        functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(32), Integer::class.java) { result ->
+//            try {
+//                assertEquals(32, result.orThrow.toInt())
+//            } finally {
+//                looperThread.testComplete()
+//            }
+//        }
+//    }
+//
+//
+//    @Test
+//    fun codecArgumentFailure() {
```

Comment:
I think it still makes sense to have some tests that verifies the exception that we throw when we cannot encode arguments and decode return values. Both when just passing some unsupported types, but also if just asking for the wrong return type, etc.

Reply:
And also that how we handle `T`/`KType` mismatches if we decide to preserve the ability to pass both in the public API

Reply:
These cases are being validated in `BsonEncoderTests`. I am going to add some tests to validate that we capture these cases here too.